Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageLoader/TIFFLoader.cpp

```diff
@@ -137,389 +154,303 @@ namespace ImageProcessingAtom
                 {
-                    for (uint32_t dwX = 0; dwX < dwWidth; ++dwX)
-                    {
-                        dstLine[0] = srcLine[0];
-                        dstLine[1] = srcLine[1];
-                        dstLine[2] = srcLine[2];
-                        dstLine[3] = (dwChannels == 3) ? 0xFFFF : srcLine[3];
+                    isGeoTIFF = true;
 
-                        dstLine += 4;
-                        srcLine += dwChannels;
+                    // if there's an xyz scale, and the Z scale isn't 0, let's use it.
+                    if ((tagCount == 3) && (pixelScales != NULL) && (pixelScales[2] != 0.0f))
+                    {
+                        pixelValueScale = static_cast<float>(pixelScales[2]);
                     }
                 }
             }
 
-            return pRet;
-        }
+            // Track min/max values for GeoTIFFs so that we can scale the values into the 0-1 range.
+            float minChannelValue = AZStd::numeric_limits<float>::max();
+            float maxChannelValue = AZStd::numeric_limits<float>::lowest();
 
-        static IImageObject* Load16BitHDRImageFromTIFF(TIFF* tif)
-        {
-            uint32_t dwChannels = 0;
-            uint32_t dwPhotometric = 0;
-            TIFFGetField(tif, TIFFTAG_SAMPLESPERPIXEL, &dwChannels);
-            TIFFGetField(tif, TIFFTAG_PHOTOMETRIC, &dwPhotometric);
-
-            uint32_t dwWidth = 0;
-            uint32_t dwHeight = 0;
-            TIFFGetField(tif, TIFFTAG_IMAGEWIDTH, &dwWidth);
-            TIFFGetField(tif, TIFFTAG_IMAGELENGTH, &dwHeight);
-
-            EPixelFormat eFormat = ePixelFormat_R16G16B16A16F;
-
-            if (dwChannels == 1 && dwPhotometric != PHOTOMETRIC_MINISBLACK)
+            // Copy one channel of one pixel from source to destination, and optionally invert the value.
+            auto CopyPixelChannel = [bitsPerChannel, pixelValueScale, &buf, &dst, &minChannelValue, &maxChannelValue]
+                (uint32_t destPixelChannelIndex, uint32_t srcPixelChannelIndex, bool invert = false)
             {
-                eFormat = ePixelFormat_R16F;
-            }
-
-            IImageObject* pRet = IImageObject::CreateImage(dwWidth, dwHeight, 1, eFormat);
+                if (bitsPerChannel == 8)
+                {
+                    dst[destPixelChannelIndex] = invert ? (0xFF - buf[srcPixelChannelIndex]) : buf[srcPixelChannelIndex];
+                }
+                else if (bitsPerChannel == 16)
+                {
+                    // Alias the base pointers to a 16-bit channel type, then use the channel index to get to the correct pixel & channel
+                    uint16_t* buf16 = reinterpret_cast<uint16_t*>(buf.data());
+                    uint16_t* dst16 = reinterpret_cast<uint16_t*>(dst);
+                    dst16[destPixelChannelIndex] = invert ? (0xFFFF - buf16[srcPixelChannelIndex]) : buf16[srcPixelChannelIndex];
+                }
+                else
```

Comment:
It looks the 16bit float input isn't handled here which is supported in GetOutputPixelFormat() as well as old implementation.

Reply:
I believe this should work for 16-bit float inputs identically to how it did before. Ignoring the invert piece for a second (which only applies to 2-channel PHOTOMETRIC_SEPARATED TIFFs), it's just doing a 16-bit copy, so any 16-bit float values should be preserved.

For reference, the previous implementation in Load16BitHDRImageFromTIFF looked like this (some of it is removed for brevity / readability):
```
            static const uint16_t zero = 0;
            static const uint16_t one = 1;

            for (uint32_t dwY = 0; dwY < dwHeight; ++dwY)
            {
                TIFFReadScanline(tif, &buf[0], dwY, 0); // read raw row

                const uint16_t* srcLine = (const uint16_t*)&buf[0];
                uint16_t* dstLine = (uint16_t*)&dst[dwPitch * dwY];

                if (dwChannels == 1)
                {
                        for (uint32_t dwX = 0; dwX < dwWidth; ++dwX)
                        {
                            dstLine[0] = srcLine[0];
                        }
                }
                else if (dwChannels == 2)
                {
                    if (dwPhotometric == PHOTOMETRIC_SEPARATED)
                    {
                        for (uint32_t dwX = 0; dwX < dwWidth; ++dwX)
                        {
                            //but convert CMY to RGB (PHOTOMETRIC_SEPARATED refers to inks in TIFF, the value is inverted)
                            dstLine[0] = uint16_t(1.0f - srcLine[0]);
                            dstLine[1] = uint16_t(1.0f - srcLine[1]);
                            dstLine[2] = zero;
                            dstLine[3] = one;
                        }
                    }
                }
                else
                {
                    for (uint32_t dwX = 0; dwX < dwWidth; ++dwX)
                    {
                        dstLine[0] = srcLine[0];
                        dstLine[1] = srcLine[1];
                        dstLine[2] = srcLine[2];
                        dstLine[3] = (dwChannels == 3) ? one : srcLine[3];
                    }
                }
            }
```
Ignoring the 2-channel PHOTOMETRIC_SEPARATED, it basically just copied the data using uint16 pointers, same as the new implementation.

I'm fairly certain that both the previous and current PHOTOMETRIC_SEPARATED implementation is broken for 16-bit floats, but I don't have any valid test data for this, so I wasn't going to try and fix it. The previous logic was doing "uint16(1.0f - srcLine[x])", where srcLine[x] is a uint16_t, so I'm skeptical that this produced mathematically valid half-float results. The current logic is instead inverting with effectively "0xFFFF - srcLine[x]", which is probably *also* invalid, but in a different way.